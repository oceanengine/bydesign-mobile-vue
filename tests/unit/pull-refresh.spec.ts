import { mount, triggerDrag, trigger, later } from '../index';
import { PullRefresh as NativePullRefresh } from '@src'
const PullRefresh = NativePullRefresh as any;
triggerDrag;
later;
describe('PullRefresh', () => {
  test('change head content when pulling down', async () => {
    const wrapper = mount({
      template: `
        <byted-pull-refresh
          v-model="isLoading"
          foot-disabled
          :load-end-status="loadEndStatus"
          @pull-down="pullDown">
          下拉刷新
        </byted-pull-refresh>
      `,
      components: {
        'byted-pull-refresh': PullRefresh,
      },
      data() {
        return {
          isLoading: false,
          loadEndStatus: '',
        };
      },
      methods: {
        pullDown() {
          wrapper.setData({ loadEndStatus: 'success', isLoading: false });
        },
      },
    });
    expect(wrapper).toMatchSnapshot();
    const track = wrapper.find('.byted-pull-refresh__track');
    // pulling
    trigger(track, 'touchstart', 0, 0);
    trigger(track, 'touchmove', 0, 20);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loading
    trigger(track, 'touchmove', 0, 100);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loading
    trigger(track, 'touchend', 0, 100);

    await later();

    // console.log(wrapper.vm['isLoading'], 'end');
    expect(wrapper).toMatchSnapshot();

    // end loading
    await later(500);
    expect(wrapper).toMatchSnapshot();
  });

  test('change head content when pulling up', async () => {
    const wrapper = mount({
      template: `
        <byted-pull-refresh
          v-model="isLoading"
          headDisabled
          :load-end-status="loadEndStatus"
          @pull-up="pullUp">
          上拉加载
        </byted-pull-refresh>
      `,
      components: {
        'byted-pull-refresh': PullRefresh,
      },
      data() {
        return {
          isLoading: false,
          loadEndStatus: '',
        };
      },
      methods: {
        pullUp() {
          wrapper.setData({ loadEndStatus: 'success', isLoading: false });
        },
      },
    });
    expect(wrapper).toMatchSnapshot();
    const track = wrapper.find('.byted-pull-refresh__track');
    // pulling
    trigger(track, 'touchstart', 0, 0);
    trigger(track, 'touchmove', 0, -20);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loading
    trigger(track, 'touchmove', 0, -100);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loading
    trigger(track, 'touchend', 0, -100);

    await later();

    // console.log(wrapper.vm['isLoading'], 'end');
    expect(wrapper).toMatchSnapshot();

    // end loading
    await later(500);
    expect(wrapper).toMatchSnapshot();
  });
  test('custom content by slots', async () => {
    const wrapper = mount(PullRefresh, {
      propsData: {
        loadEndStatus: '',
        value: false,
        footDisabled: true,
      },
      scopedSlots: {
        pulling({ distance }) {
          return `pulling ${distance}`;
        },
        loosing({ distance }) {
          return `loosing ${distance}`;
        },
        loading({ distance }) {
          return `loading ${distance}`;
        },
      },
    });

    const track = wrapper.find('.byted-pull-refresh__track');

    // pulling
    trigger(track, 'touchstart', 0, 0);
    trigger(track, 'touchmove', 0, 20);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loosing
    trigger(track, 'touchmove', 0, 75);
    trigger(track, 'touchmove', 0, 100);
    await later();
    expect(wrapper).toMatchSnapshot();

    // loading
    trigger(track, 'touchend', 0, 100);
    await later();
    expect(wrapper).toMatchSnapshot();
  });

  test('pull a short distance', () => {
    const wrapper = mount(PullRefresh, {
      propsData: {
        loadEndStatus: '',
        value: false,
        footDisabled: true,
      },
      listeners: {
        pullDown() {
          console.log('00000');
        },
      },
    });

    const track = wrapper.find('.byted-pull-refresh__track');
    triggerDrag(track, 0, 10);
    expect(wrapper.emitted('input')).toBeFalsy();
  });
});
