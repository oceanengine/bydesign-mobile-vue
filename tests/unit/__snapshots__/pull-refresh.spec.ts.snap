// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PullRefresh change head content when pulling down 1`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms;">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    下拉刷新
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling down 2`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,20px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--pulling"><span class="byted-pull-refresh__wrp__text">下拉刷新</span></div>
    </div>
    下拉刷新
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling down 3`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,75px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--loosing"><span class="byted-pull-refresh__wrp__text">释放刷新</span></div>
    </div>
    下拉刷新
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling down 4`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 300ms; transform: translate3d(0,50px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--success"><span class="byted-pull-refresh__wrp__text">刷新成功</span></div>
    </div>
    下拉刷新
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling down 5`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 300ms;">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    下拉刷新
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling up 1`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms;">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    上拉加载
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling up 2`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,-20px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    上拉加载
    <div class="byted-pull-refresh__foot" style="bottom: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--pulling"><span class="byted-pull-refresh__wrp__text">下拉刷新</span></div>
    </div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling up 3`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,-75px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    上拉加载
    <div class="byted-pull-refresh__foot" style="bottom: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--loosing"><span class="byted-pull-refresh__wrp__text">释放刷新</span></div>
    </div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling up 4`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 300ms; transform: translate3d(0,-50px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    上拉加载
    <div class="byted-pull-refresh__foot" style="bottom: -34px;">
      <div class="byted-pull-refresh__wrp byted-pull-refresh__wrp--default byted-pull-refresh__wrp--normal byted-pull-refresh__wrp--success"><span class="byted-pull-refresh__wrp__text">刷新成功</span></div>
    </div>
  </div>
</div>
`;

exports[`PullRefresh change head content when pulling up 5`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 300ms;">
    <div class="byted-pull-refresh__head" style="top: -34px;"></div>
    上拉加载
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh custom content by slots 1`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,20px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">pulling 20</div>
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh custom content by slots 2`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 0ms; transform: translate3d(0,75px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">loosing 75</div>
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;

exports[`PullRefresh custom content by slots 3`] = `
<div class="byted-pull-refresh">
  <div class="byted-pull-refresh__track" style="transition: 300ms; transform: translate3d(0,50px, 0);">
    <div class="byted-pull-refresh__head" style="top: -34px;">loading 50</div>
    <div class="byted-pull-refresh__foot" style="bottom: -34px;"></div>
  </div>
</div>
`;
